lib/async/redis/sentinels.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+module Async
+	module Redis
+		class SentinelsClient < Client
+			def initialize(master_name, sentinels, role = :master, protocol = Protocol::RESP2, **options)
+				@master_name = master_name
+				@sentinel_endpoints = sentinels.map do |sentinel|
+					Async::IO::Endpoint.tcp(sentinel[:host], sentinel[:port])
+				end
+				@role = role
+
+				@protocol = protocol
+				@pool = connect(**options)
+			end
+
+			private
+
+			# Override the parent method. The only difference is that this one needs
+			# to resolve the master/slave address.
+			def connect(**options)
+				Async::Pool::Controller.wrap(**options) do
+					address = resolve_address
+					endpoint = Async::IO::Endpoint.tcp(address[:ip], address[:port])
+
+					peer = endpoint.connect
+					stream = IO::Stream.new(peer)
+
+					@protocol.client(stream)
+				end
+			end
+
+			def resolve_address
+				address = case @role
+				when :master
+					resolve_master
+				when :slave
+					resolve_slave
+				else
+					raise ArgumentError, "Unknown instance role #{@role}"
+				end
+
+				address or raise RuntimeError, "Unable to fetch #{@role} via Sentinel."
+			end
+
+			def resolve_master
+				@sentinel_endpoints.each do |sentinel_endpoint|
+					client = Client.new(sentinel_endpoint)
+
+					begin
+						address = client.call('sentinel', 'get-master-addr-by-name', @master_name)
+					rescue Errno::ECONNREFUSED
+						next
+					end
+
+					return { host: address[0], port: address[1] } if address
+				end
+
+				nil
+			end
+
+			def resolve_slave
+				@sentinel_endpoints.each do |sentinel_endpoint|
+					client = Client.new(sentinel_endpoint)
+
+					begin
+						reply = client.call('sentinel', 'slaves', @master_name)
+					rescue Errno::ECONNREFUSED
+						next
+					end
+
+					if slave = select_slave(reply)
+						return { host: slave['ip'], port: slave['port'] }
+					end
+				end
+
+				nil
+			end
+
+			def select_slave(slaves_cmd_reply)
+				return nil unless slaves_cmd_reply
```

Comment:
If a user did decide to implement their own policy, they would need to parse `slaves_cmd_reply` themselves.

I think we should parse it in `resolve_slave` either directly, or using a helper, into a data structure which you can easily make sense of (e.g. the hash that you construct already).

Then, `select_slave(available)` would just be `available.sample`. Some other implementation might prefer more complex logic (like checking latency).

Reply:
That's a good point. I've separated the two methods in a new commit https://github.com/socketry/async-redis/pull/29/commits/3c6c5ae281be35eded9df7c11f7ea0c721a8fb02
I think it's better now.